pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1450,65 +1058,194 @@ char *Series_iterGetName(JSOBJ obj, JSONTypeContext *tc, size_t *outLen) {
-        if (is_simple_frame(obj)) {
-            GET_TC(tc)->itemValue = get_values(obj);
-            if (!GET_TC(tc)->itemValue) {
-                return 0;
-            }
-        } else {
-            Py_INCREF(obj);
-            GET_TC(tc)->itemValue = obj;
-        }
-    } else {
+	GET_TC(tc)->itemValue = get_values(obj);
+	if (!GET_TC(tc)->itemValue) {
+	  return 0;
+	}
+      } else {
         PRINTMARK();
         return 0;
+      }
+
+    } else {
+      if (enc->originalOutputFormat == COLUMNS || enc->originalOutputFormat == INDEX) {
+	// TODO: align extension usage of itemName and cStr
+	if (GET_TC(tc)->itemName) {
+	  Py_DECREF(GET_TC(tc)->itemName);
+	}
+      }
+
+      PyObject *tmp = PyIter_Next(GET_TC(tc)->iterator);
+      if (tmp == 0)
+	return 0;
+
+      GET_TC(tc)->itemValue = PySequence_GetItem(tmp, 1);
+
+      if (enc->originalOutputFormat == COLUMNS || enc->originalOutputFormat == INDEX) {
+	GET_TC(tc)->itemName = PySequence_GetItem(tmp, 0);
+      }
+
+      Py_DECREF(tmp);
     }
 
     GET_TC(tc)->index++;
     PRINTMARK();
     return 1;
 }
 
+/*
+ * Function: DataFrame_iterEnd
+ * -----------------------------
+ * Callaback after DataFrame has been entirely iterated upon.
+ *
+ */
 void DataFrame_iterEnd(JSOBJ obj, JSONTypeContext *tc) {
-    PyObjectEncoder *enc = (PyObjectEncoder *)tc->encoder;
-    enc->outputFormat = enc->originalOutputFormat;
-    PRINTMARK();
+  PRINTMARK();
+  PyObjectEncoder *enc = (PyObjectEncoder *)tc->encoder;
+
+  if (enc->originalOutputFormat != SPLIT) {
+    Py_DECREF(GET_TC(tc)->iterator);
+  }
+
+  enc->outputFormat = enc->originalOutputFormat;
+
 }
 
+/*
+ * Function: DataFrame_iterGetValue
+ * -----------------------------
+ * Provides the value(s) for a particular iteration. This is valid whether
+ * the type context is JT_OBJECT or JT_ARRAY.
+ */
 JSOBJ DataFrame_iterGetValue(JSOBJ obj, JSONTypeContext *tc) {
-    return GET_TC(tc)->itemValue;
+  PRINTMARK();
+  return GET_TC(tc)->itemValue;
 }
 
+/*
+ * Function: DataFrame_iterGetName
+ * -----------------------------
+ * Provides the name for a particular iteration. This is only called if
+ * the type context is JT_OBJECT, which is dictated by the orient.
+ */
 char *DataFrame_iterGetName(JSOBJ obj, JSONTypeContext *tc, size_t *outLen) {
+  PRINTMARK();
+  PyObjectEncoder *enc = (PyObjectEncoder *)tc->encoder;
+  if (enc->originalOutputFormat == SPLIT) {
     *outLen = strlen(GET_TC(tc)->cStr);
     return GET_TC(tc)->cStr;
+  } else if (enc->originalOutputFormat == COLUMNS || enc->originalOutputFormat == INDEX) {
+    // TODO: index is incremented before iteration completes which is unfortunate
+    // Need to align with SPLIT format and then can maybe increment here for clarity
+    int index = GET_TC(tc)->index - 1;
+
+    // Also TODO: return a value here rather than having NpyArr_getLabel modify output buf
+    NpyArr_getLabel(obj, tc, outLen, index, GET_TC(tc)->columnLabels);
```

Comment:
It took me a while to grok the use of this function and I find it rather unreadable but leaving for a subsequent PR. With ujson key / value pairs you can override the method to provide the key for a particular object, which would be the idiomatic approach.

I guess as a result of blocks not mapping cleanly to key / value pairs the historical approach was to encode the labels separately and use this dedicated method to write directly to the output buffer rather than returning a `char *` for ujson to encode into the buffer automatically. 

It would take a rather large effort so leaving to a separate PR but would like to use ujson iteration instead of manually encoding all values up front